demo/src/error.rs

```diff
@@ -1,52 +1,86 @@
 use std::io;
-use termion::event::Key;
+use thiserror;
 
 use editor::DocError;
-use frontends::terminal;
+use frontends::{terminal::TermError, Key};
 use language::{ConstructName, LanguageName};
-use pretty::PaneError;
+use pretty::{DocLabel, PaneError};
 
-#[derive(Debug)]
-pub enum Error {
+use crate::keymaps::{MenuName, ModeName};
+
+#[derive(thiserror::Error, Debug)]
+pub enum ServerError<'l> {
+    #[error("not in keymap: {0:?}")]
     UnknownKey(Key),
-    UnknownKeymap(String),
-    NoKeymap,
+
+    #[error("unknown keymap mode: {0:?}")]
+    UnknownModeName(ModeName),
+
+    #[error("unknown keymap menu: {0:?}")]
+    UnknownMenuName(MenuName),
+
+    #[error("no keymap mode selected")]
+    NoMode,
+
+    #[error("unknown user input event")]
     UnknownEvent,
+
+    #[error("received keyboard interrupt")]
     KeyboardInterrupt,
+
+    // TODO include actual type too
+    #[error("expected value of type {0} on data stack")]
+    ExpectedValue(String),
+
+    #[error("data stack was unexpectedly empty")]
+    EmptyDataStack,
+
+    #[error("input/output error: {0}")]
+    Io(#[from] io::Error),
+
+    #[error("terminal error: {0}")]
+    Term(#[from] TermError),
+
+    // Note: we can't use the inner EngineError as the error source because it
+    // contains a non-static lifetime.
+    #[error("engine error: {0}")]
+    Engine(EngineError<'l>),
+}
+
+impl<'l> From<EngineError<'l>> for ServerError<'l> {
+    fn from(e: EngineError<'l>) -> ServerError<'l> {
+        ServerError::Engine(e)
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum EngineError<'l> {
+    #[error("unknown language: {0}")]
     UnknownLang(LanguageName),
+
+    #[error("language {lang} does not contain construct {construct}")]
     UnknownConstruct {
         construct: ConstructName,
         lang: LanguageName,
     },
+
+    #[error("no bookmark stored with this key")]
     UnknownBookmark,
-    ExpectedWord(String),
-    EmptyStack,
-    Pane(PaneError<terminal::Error>),
-    DocExec(DocError<'static>),
-    Io(io::Error),
-    Term(terminal::Error),
-}
 
-impl From<PaneError<terminal::Error>> for Error {
-    fn from(e: PaneError<terminal::Error>) -> Error {
-        Error::Pane(e)
-    }
-}
+    #[error("no document with label {0:?}")]
+    UnknownDocLabel(DocLabel),
 
-impl From<io::Error> for Error {
-    fn from(e: io::Error) -> Error {
-        Error::Io(e)
-    }
-}
+    #[error("pane error: {0}")]
+    Pane(#[from] PaneError),
 
-impl From<terminal::Error> for Error {
-    fn from(e: terminal::Error) -> Error {
-        Error::Term(e)
-    }
+    // Note: we can't use the inner DocError as the error source because it
+    // contains a non-static lifetime.
+    #[error("doc error: {0}")]
+    DocExec(DocError<'l>),
```

Comment:
Removed the lifetime parameter from errors. We need to keep the `Replace` command though, for implementing undo. Because of that, and because the Doc doesn't have access to the language set and notation sets needed to create a new node, I'd rather continue to have the Engine create new nodes and pass them to the Doc in a `Replace` command. The Doc won't be able to give back a rejected node to the Engine, but that doesn't matter.